corehq/warehouse/models/shared.py

```diff
@@ -0,0 +1,86 @@
+import os
+
+from django.db import connections
+from django.conf import settings
+from django.template import Context, engines
+
+from corehq.sql_db.routers import db_for_read_write
+
+from corehq.warehouse.transforms import get_transform
+from corehq.warehouse.const import (
+    DJANGO_MAX_BATCH_SIZE,
+)
+
+
+class WarehouseTableMixin(object):
+
+    @classmethod
+    def sql_query_template_name(cls):
+        return None
+
+    @classmethod
+    def dependencies(cls):
+        raise NotImplementedError
+
+    @classmethod
+    def load(cls):
```

Comment:
we should avoid updating/inserting one row at a time. we can actually load the dim in just a few bulk queries using an extra intermediate table which has the same exact structure as the final dim and also includes one extra boolean column: is_new, which is set to true by default.  the process would be:

1) truncate intermediate table and insert data from staging table into intermediate table, using an insert ... select statement, and perform any transformations along with the select (e.g., doc_type -> deleted flag)

2) update the intermediate table using an update ... from statement, joining with the dim on doc id, and setting the is_new flag to false for anything that matches

3) update the dim using an update ... from statement, joining with the intermediate table on doc id

4) insert into the dim selecting from the intermediate table where is_new is true

5) vacuum analyze the dim (this should be the last step of every dim / fact loading process)

this should suffice for almost all our dims. how we load our facts will be dependent on the complexity of the fact table, and each one might be different